utility/servicer/module.go

```diff
@@ -182,31 +356,107 @@ func (s servicer) admitRelay(relay *coreTypes.Relay) error {
+		return fmt.Errorf("%s: %w", err.Error(), errValidateRelayMeta)
 	}
 
 	// TODO: update the CLI to include ApplicationAddress(or Application Public Key) in the RelayMeta
-	session, err := s.GetBus().GetUtilityModule().GetSession(relay.Meta.ApplicationAddress, int64(height), relay.Meta.RelayChain.Id, relay.Meta.GeoZone.Id)
+	session, err := s.getSession(relay)
 	if err != nil {
-		return fmt.Errorf("%s: failed to get a session for height %d for relay meta %s: %w", errPrefix, height, relay.Meta, err)
+		return err
 	}
 
 	// TODO: (REFACTOR) use a loop to run all validators: would also remove the need for passing the session around
 	if err := validateRelayBlockHeight(relay.Meta, session); err != nil {
-		return fmt.Errorf("%w: %s", errValidateBlockHeight, err.Error())
+		return fmt.Errorf("%s: %w", err.Error(), errValidateBlockHeight)
 	}
 
-	if err := s.validateApplication(relay.Meta, session); err != nil {
-		return fmt.Errorf("%s: relay failed application validation: %w", errPrefix, err)
+	if err := s.validateServicer(relay.Meta, session); err != nil {
+		return fmt.Errorf("%s: %s: %w", errPrefix, err.Error(), errValidateServicer)
 	}
 
-	if err := s.validateServicer(relay.Meta, session); err != nil {
-		return fmt.Errorf("%s: relay failed servicer instance validation: %w", errPrefix, err)
+	if err := s.validateApplication(session); err != nil {
+		return fmt.Errorf("%s: %s: %w", errPrefix, err.Error(), errValidateApplication)
 	}
 
 	return nil
 }
 
+// DISCUSS: do we need to export this functionality as part of the utility module?
+// calculateAppSessionTokens determines the number of "session tokens" an application gets at the beginning
+// of every session. Each servicer will serve a maximum of ~(Session Tokens / Number of Servicers in the Session) relays for the application
+func (s *servicer) calculateAppSessionTokens(session *coreTypes.Session) (*big.Int, error) {
+	appStake, err := utils.StringToBigInt(session.Application.StakedAmount)
+	if err != nil {
+		return nil, fmt.Errorf("Error processing application's staked amount %s: %w", session.Application.StakedAmount, coreTypes.ErrStringToBigInt(err))
+	}
+
+	// DOCUMENT: find the right document to explain the following:
+	//	We assume that the value of certain parameters only changes/takes effect at the start of a session.
+	//	In this specific case, the `AppSessionTokensMultiplierParamName` parameter is retrieved for the height that
+	//		matches the beginning of the session.
+	readCtx, err := s.GetBus().GetPersistenceModule().NewReadContext(session.SessionHeight)
+	if err != nil {
+		return nil, fmt.Errorf("error getting persistence context at height %d: %w", session.SessionHeight, err)
+	}
+	defer readCtx.Release() //nolint:errcheck // We only need to make sure the readCtx is released
+
+	// DISCUSS: using an interface for returning each defined parameter seems less error-prone: also could return e.g. int64 in this case to remove the type cast
```

Comment:
Have you looked at the `GetParameter` function?

Reply:
Thanks, that solved it. Removed the `DISCUSS` item.